rest/http_client.go

```diff
@@ -20,21 +21,38 @@ type HTTPClient interface {
 	Close()
 	Logger() log.Logger
 	HTTPClient() *http.Client
+	RateLimiter() rate.RateLimiter
 	UserAgent() string
 	Do(ctx context.Context, route *route.CompiledAPIRoute, rqBody interface{}, rsBody interface{}) Error
-	DoWithHeaders(ctx context.Context, route *route.CompiledAPIRoute, rqBody interface{}, rsBody interface{}, customHeader http.Header) Error
 }
 
 // NewHTTPClient constructs a new HTTPClient with the given http.Client, log.Logger & useragent
 //goland:noinspection GoUnusedExportedFunction
-func NewHTTPClient(logger log.Logger, httpClient *http.Client, userAgent string) HTTPClient {
-	return &HTTPClientImpl{userAgent: userAgent, httpClient: httpClient, logger: logger}
+func NewHTTPClient(logger log.Logger, httpClient *http.Client, rateLimiter rate.RateLimiter, headers http.Header, userAgent string) HTTPClient {
+	if logger == nil {
+		logger = log.Default()
+	}
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+	if rateLimiter == nil {
+		rateLimiter = rate.NewRateLimiter(logger)
```

Comment:
would it be worth having a RateLimiterConfig to configure the behaviour of the ratelimiter? eg; handling retries and what should happen on a 429>

Reply:
My thought was putting that into the http client as that also handles the retry itself
But yes we should

Reply:
nvm I ended up putting it in the RateLimiter Config haha